### PR TITLE
Improve benchmark tools

### PR DESCRIPTION
**Pull request by d-ronnqvist**
Bug/issue #, if applicable: rdar://87993211

## Summary

This improves the benchmark tools in a few ways:
- All benchmark files are easily decodable, doesn't discard data, sorted for presentation, and encodes the type of metric values so it doesn't need to be parsed out of the description of the unit in the metric display name.
- The diff tool makes an assessment of the statistical significance of the difference and reports indicators of invalid diff results.
- The benchmark code is reorganized into a local Swift package that makes it easier to share code between benchmark tools
- A new `compare-to` tool makes it more convenient to diff changes on a branch to another commit (for example the current HEAD of the main branch)

### Remark

It's been many years since I last took a statistics class so there may be mistakes in either the implementation or application of these methods. I think that I got it right and certainly feel that these results will be more trustworthy than the previous implementation. 

Since it's possible to re-measure historical commits we can always make fixes to the diff analysis over time.

## Dependencies

n/a

## Testing

To test these changes, try out the different commands. To start with

```
swift run --package-path bin/benchmark benchmark --help
```

will list the new commands and
```
swift run --package-path bin/benchmark benchmark <subcommand> --help
```

will list the arguments, flags, and options for each command.

### Updated commands

The `measure` subcommand is the new version of `bin/benchmark.swift` although the aggregate benchmark JSON format is different.

The `diff` subcommand is the new version of `bin/benchmark-diff.swift` although the analysis and the output format is different.

### New commands

The `compare-to`, `measure-commits`, and `render-trend` commands are new.

Use the `compare-to` to compare the docc performance of this branch against another "commit-ish" (for example `df272bf89f01e807d30b5e50fd86fa8b8a8ebbc4`). The diff analysis shouldn't show any significant differences since the docc code is the same (except for minor differences in the benchmark encoding).

Use the `measure-commits` to gather docc benchmark measurements for two or more commits. Then use the `diff` command to compare the benchmark results from the different commits.

Use the `render-trend` to visualize trends in the numerical metrics for the benchmark measurements that you gathered from the `measure-commits` command. This visualization will make more sense if you gather more measurements for more than two commits. Optionally, pass filters to limit what metrics are visualized.

## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.

- [x] Added tests
- [x] Ran the `./bin/test` script and it succeeded
- [x] Updated documentation if necessary

### ~Remaining~

~This is mostly done but I have a few tasks left that I intend to do:~

- [x] ~Update the conceptual documentation in `Benchmarking.md` and `benchmark.md`~
- [x] ~Add an option to fill in output size measurements that's missing in older checkouts~
- [x] ~Add some form of basic visualization of trends across multiple benchmarks.~
- [x] ~Provide testing instructions (which has some overlap with updating the conceptual documentation)~
- [x] ~Fix argument parsing bug in `measure-commits` tool~
- [x] ~Add option to write diff results to a JSON file so that scripts don't need to parse the table~
- [x] ~Update benchmark.json OpenAPI Spec~

### CODE REVIEW

**Comment by d-ronnqvist** [comment]
@swift-ci please test

**Comment by d-ronnqvist** [comment]
@swift-ci please test

**Comment by d-ronnqvist** [comment]
@swift-ci please test

**Comment by d-ronnqvist** [comment]
@swift-ci please test

File: bin/benchmark/Sources/benchmark/Subcommands/Measure.swift [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+    task.waitUntilExit()
+    
+    
+    guard task.terminationStatus == 0 else {
+        throw Measure.Error.subtaskTerminated(task.terminationStatus, task.terminationReason)
+    }
+}
+
+private func waitUntilFairThermalState(minimumWait: Int) throws {
+    print("Preparing for next stage", terminator: " ")
```
Comment (franklinsch) [inline bin/benchmark/Sources/benchmark/Subcommands/Measure.swift:291]: Nit
```suggestion
    print("Preparing for next stage", terminator: " ")
```

File: bin/benchmark/Sources/benchmark/Diff/DiffResultsTable.swift [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+                    change = "no change"
+                case .differentChecksum:
+                    change = "change"
+                case .differentNumeric(let percentage):
+                    change = percentageFormatter.string(from: NSNumber(value: percentage))!
+                case .notApplicable:
+                    change = "n/a"
+            }
+            if let footnotes = analysis.footnotes, !footnotes.isEmpty {
+                let footNoteSuffix = (footnoteCounter ..< footnoteCounter+footnotes.count).map { Self.superscriptCharacters[$0] }.joined(separator: ",")
```
Comment (franklinsch) [inline bin/benchmark/Sources/benchmark/Diff/DiffResultsTable.swift:52]: Nit: the superscripts look a bit odd with the non-superscript comma separators. I'm wondering if instead of superscripts altogether, we could present the detailed information below the table by repeating the metric titles.

File: bin/benchmark/Sources/benchmark/Diff/DiffAnalysis.swift [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+                    .init(id: metric.id, displayName: metric.displayName, values: .bytesOnDisk([$0]))
+                }
+                
+            case .checksum:
+                // Only numeric metrics support trend bars
+                fatalError("Can't compute input bias for checksum metrics.")
+        }
+        
+        return """
+        The '\(metric.id)' samples from the '\(sampleName)' measurements show _possible_ signs of bias.
```
Comment (franklinsch) [inline bin/benchmark/Sources/benchmark/Diff/DiffAnalysis.swift:151]: It would be good to clearly state what this entails for the analysis in this sentence.

```suggestion
        warning: Results for the '\(metric.id)' samples for '\(sampleName)' are not normally distributed. Run the benchmark again or inspect the results manually.
```

File: bin/benchmark/Sources/benchmark/Subcommands/RenderTrend.swift [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+                // Skip filtered-out metrics
+                continue
+            }
+            
+            print(metrics[0].displayName)
+            print(Self.renderTrendBars(metrics: metrics))
+        }
+    }
+
+    static func renderTrendBars(metrics: [BenchmarkResultSeries.MetricSeries]) -> String {
```
Comment (franklinsch) [inline bin/benchmark/Sources/benchmark/Subcommands/RenderTrend.swift:79]: This is a great idea, but the bars sometimes look a bit jarring. I'm wondering if displaying "|" characters would provide the same data in a simpler way.

Screen Shot 2022-05-04 at 14 13 46

File: bin/benchmark/Sources/benchmark/Subcommands/Measure.swift [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+        
+        // This adds some extra wait time before the first measurement.
+        try waitUntilFairThermalState(minimumWait: 5)
+        
+        var benchmarkSeries = BenchmarkResultSeries.empty
+        
+        for conversionIndex in 1...repeatCount {
+            try waitUntilFairThermalState(minimumWait: 10)
+            
+            print("Measuring data for sample [\(conversionIndex) / \(repeatCount)]")
```
Comment (franklinsch) [inline bin/benchmark/Sources/benchmark/Subcommands/Measure.swift:182]: It's hard to differentiate logs from the `docc` invocation vs. logs from the benchmarking tool. I'm wondering if we should show benchmark logs in a different color, or just drop the `docc` logs.

File: bin/benchmark/Sources/benchmark/Diff/DiffAnalysis.swift [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+        The benchmark diff analysis assumes that these values follow a normal distribution around the value that they're meant to represent. If they don't then the conclusion from the diff analysis could be invalid.
+        A human should inspect these values. If the samples look biased or look too noisy you can try gathering new metrics with more 'repetitions' or with a larger .docc catalog.
+        [\(numbers.map { warningNumberFormatter.string(from: NSNumber(value: $0))! }.joined(separator: ", "))]
+        \(RenderTrendAction.renderTrendBars(metrics: metricSeries))
```
Comment (franklinsch) [inline bin/benchmark/Sources/benchmark/Diff/DiffAnalysis.swift:155]: I'm not sure sure I understand why the numbers displayed in the array above are different to the ones in the trend bars. Is there a reason for that?

File: bin/benchmark/Sources/benchmark/Diff/DiffAnalysis.swift [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+            change = .same
+            footnotes = [
+                .init(text: "The before and after values are similar enough that the most probable explanation is that they're random samples from the same data set.", values: footnoteValues)
+            ]
+        } else {
+            let approximateChange = (afterNumbers.mean() - beforeNumbers.mean()) / beforeNumbers.mean()
+            change = .differentNumeric(percentage: approximateChange)
+            
+            footnotes = [
+                .init(text: "The before and after values are different enough that the most probable explanation is that random samples from two different data sets.", values: footnoteValues)
```
Comment (franklinsch) [inline bin/benchmark/Sources/benchmark/Diff/DiffAnalysis.swift:105]: I think there is a missing word here.

Also, I find this text and the text above a bit hard to follow, not having a statistics background.

Is it correct to simply say:
- ✅ No statistically significant change
- ⚠️ Possible regression

And display that in the table, so that you can see at a glance what might've regressed?

File: bin/benchmark/Sources/benchmark/Subcommands/CommitCommands.swift [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+ See https://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import Foundation
+import ArgumentParser
+import SwiftDocC
+
+struct CompareTo: ParsableCommand {
+    @Argument(
+        help: "The baseline 'commit-ish' to compare the current checkout against."
```
Comment (franklinsch) [inline bin/benchmark/Sources/benchmark/Subcommands/CommitCommands.swift:17]: Instead of 'commit-ish', is 'revision' accurate?

File: bin/benchmark/Sources/benchmark/Subcommands/CommitCommands.swift [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+    )
+    var commitHash: String
```
Comment (franklinsch) [inline bin/benchmark/Sources/benchmark/Subcommands/CommitCommands.swift:19]: As an extension to this, it might be useful to provide an argument that takes a path to another clone of `docc`, so that you can run comparisons against forks as well.

File: bin/benchmark/Sources/benchmark/Subcommands/Diff.swift [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+        DiffResultsTable.columns[3] = tableColumnInfo(file: afterFile)
+        
+        let table = DiffResultsTable(results: result)
+        print(table.output)
+        
+        if let jsonOutputFile = jsonOutputFile {
+            let encoder = JSONEncoder()
+            encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
+            try encoder.encode(result).write(to: jsonOutputFile)
+        }
```
Comment (franklinsch) [inline bin/benchmark/Sources/benchmark/Subcommands/Diff.swift:75]: Should we exit with a non-zero code if one of the comparisons led to a statistically significant result? That would make it easier to spot in CI whether a change introduced a regression.

**Review by franklinsch — COMMENTED** [review]
This looking incredible, thank you for doing this! Just a few suggestions for your consideration

File: bin/benchmark/Sources/benchmark/Subcommands/Measure.swift (hunk 1 above)
Comment (d-ronnqvist) [reply bin/benchmark/Sources/benchmark/Subcommands/Measure.swift:291]: I can move this space to the end of the string instead of the terminator.

File: bin/benchmark/Sources/benchmark/Diff/DiffAnalysis.swift (hunk 7 above)
Comment (d-ronnqvist) [reply bin/benchmark/Sources/benchmark/Diff/DiffAnalysis.swift:105]: Franklin and I talked about this offline. I feel that the diff tool should avoid trying to draw conclusions and instead aim to present the necessary information that a human needs to consider when analyzing the performance differences.

To that end, I feel like we should be careful to avoid present inconclusive results that may be incorrectly interpreted as absolute success/failure values. Still, we could use colors or formatting to guide the person who looks at the results to the values that they need to focus their attention on.

File: bin/benchmark/Sources/benchmark/Subcommands/Measure.swift (hunk 1 above)
Comment (d-ronnqvist) [reply bin/benchmark/Sources/benchmark/Subcommands/Measure.swift]: Addressed in c34eaaa078ea63733969ca2e21f5449bf7d336ef

File: bin/benchmark/Sources/benchmark/Subcommands/Measure.swift (hunk 5 above)
Comment (d-ronnqvist) [reply bin/benchmark/Sources/benchmark/Subcommands/Measure.swift]: I addressed this in c34eaaa078ea63733969ca2e21f5449bf7d336ef by bolding the benchmark output

File: bin/benchmark/Sources/benchmark/Diff/DiffAnalysis.swift (hunk 7 above)
Comment (d-ronnqvist) [reply bin/benchmark/Sources/benchmark/Diff/DiffAnalysis.swift]: I addressed this in c34eaaa078ea63733969ca2e21f5449bf7d336ef by adding color as follows:
- green: statistically significant improvement in performance _and_ no signs of bias in these measurements
- red: statistically significant regression in performance _and_ signs of bias in these measurements
- yellow: possible signs of signs of bias in either the before or after measurements

File: bin/benchmark/Sources/benchmark/Subcommands/CommitCommands.swift (hunk 9 above)
Comment (d-ronnqvist) [reply bin/benchmark/Sources/benchmark/Subcommands/CommitCommands.swift:19]: Can you give examples of when we'd want to compare across forks (other than comparing from a PR back to main)?

It's always possible to locally checkout a certain branch from another remote and benchmark against the commit that way. I'm not sure that comparing against forks is common enough that it would be worth automating that step.

File: bin/benchmark/Sources/benchmark/Diff/DiffAnalysis.swift (hunk 6 above)
Comment (d-ronnqvist) [reply bin/benchmark/Sources/benchmark/Diff/DiffAnalysis.swift:161]: I addressed this in c34eaaa078ea63733969ca2e21f5449bf7d336ef by printing "full precision values: " before the numerical values and dimming that line to deemphasize it.

**Comment by d-ronnqvist** [comment]
@swift-ci please test

**Review by franklinsch — APPROVED** [review]

**Comment by d-ronnqvist** [comment]
A few DocC developers have been using both versions of the benchmarking tools for a while now and are happy enough with the results to merge this.

We may make additional adjustments (for example to the diff output and the threshold for flagging metrics as "possibly biased") in follow-up PRs after it's been used by us and other developers.
